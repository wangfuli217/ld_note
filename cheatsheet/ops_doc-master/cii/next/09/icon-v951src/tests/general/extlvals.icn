#  Test use of external values with dynamic loading, using demo cfunc.

link cfunc

record complex(r, i)

procedure main()
   local e, l, v1, v2, v3, v4
   local r1,r2,r3,r4, s1, s2, s3, s4

   # test simple creation, type(), copy()
   xwrite("v1", v1 := extxmin())
   xwrite("v2", v2 := extxmin())
   xwrite("v3", v3 := v1)
   xwrite("v4", v4 := copy(v1))

   # test string-based external example
   xwrite("s1", s1 := extxstr("bite"))
   xwrite("s2", s2 := extxstr("the"))
   xwrite("s3", s3 := extxstr("wax"))
   xwrite("s4", s4 := extxstr("tadpole"))

   # test real-based external example
   xwrite("r1", r1 := extxreal(111.1))
   xwrite("r2", r2 := extxreal(222.2))
   xwrite("r3", r3 := extxreal(333.3))
   xwrite("r4", r4 := copy(r1))

   # test === and ~===
   teqv(v1, v3, v2)
   teqv(v1, v4, v2)
   teqv(s2, copy(s2), s3)
   teqv(r1, r1, copy(r1))
   teqv(s1, s1, s3)
   teqv(r2, r2, r4)
   teqv(s3, s3, r3)

   # test sorting
   l := [ v1, s2, r3, v4, s1, r2, complex(8,9), v3, s4, r1, v2, s3, r4 ]
   write()
   every xwrite("before", !l)
   write()
   every xwrite("sorted", !sort(l))

end


#  write label, type(xval), image(xval)

procedure xwrite(label, xval)
   write(label, ": ", type(xval), ": ",image(xval))
   return
end


#  test equivalence of id1 and id2, and nonequivalence of diff

procedure teqv(id1, id2, diff)
   if id1 ~=== id2 then
      write("nonequivalent: ", image(id1), " and ", image(id2))
   if id2 ~=== id1 then
      write("nonequivalent: ", image(id2), " and ", image(id1))
   if not (id1 === id2) then
      write("not equivalent: ", image(id1), " and ", image(id2))

   if id1 === diff then
      write("false equivalence: ", image(id1), " and ", image(diff))
   if id2 === diff then
      write("false equivalence: ", image(id2), " and ", image(diff))
   return
end
