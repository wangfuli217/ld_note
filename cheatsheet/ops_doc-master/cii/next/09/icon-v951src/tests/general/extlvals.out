v1: external: external_1(0)
v2: external: external_2(0)
v3: external: external_1(0)
v4: external: external_1(0)
s1: xstr: xstr_3(00324:bite)
s2: xstr: xstr_4(31681:the)
s3: xstr: xstr_5(35548:wax)
s4: xstr: xstr_6(46645:tadpole)
r1: xreal: xreal_7(111.1)
r2: xreal: xreal_8(222.2)
r3: xreal: xreal_9(333.3)
r4: xreal: xreal_10(111.1)

before: external: external_1(0)
before: xstr: xstr_4(31681:the)
before: xreal: xreal_9(333.3)
before: external: external_1(0)
before: xstr: xstr_3(00324:bite)
before: xreal: xreal_8(222.2)
before: complex: record complex_1(2)
before: external: external_1(0)
before: xstr: xstr_6(46645:tadpole)
before: xreal: xreal_7(111.1)
before: external: external_2(0)
before: xstr: xstr_5(35548:wax)
before: xreal: xreal_10(111.1)

sorted: complex: record complex_1(2)
sorted: external: external_1(0)
sorted: external: external_1(0)
sorted: external: external_1(0)
sorted: external: external_2(0)
sorted: xreal: xreal_7(111.1)
sorted: xreal: xreal_10(111.1)
sorted: xreal: xreal_8(222.2)
sorted: xreal: xreal_9(333.3)
sorted: xstr: xstr_3(00324:bite)
sorted: xstr: xstr_4(31681:the)
sorted: xstr: xstr_5(35548:wax)
sorted: xstr: xstr_6(46645:tadpole)
